### PR TITLE
Enable JMX Connections on a secure or insecure JMX port on Zookeeper nodes

### PR DESCRIPTION
**Pull request by OuesFa**
Signed-off-by: faresoueslati <fares.oueslati@blablacar.com>

### Type of change

- Enhancement / new feature

### Description

Created a new field in the Kafka CRD which would allow to enable the
JMX metrics on the Zookeeper nodes (similarly to Kafka brokers and Connect workers). 
When the field is specified, the Zookeeper
nodes starts up with port 9999 for the JMX. Then it exposes 9999 on the
Zookeeper headless service to be able to collect the metrics.

Enabled the ability to start up port with password and username
protection. 

### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [] Reference relevant issue(s) and close them after merging
- [x] Update CHANGELOG.md
- [] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

**Comment by OuesFa** [comment]
If I'm not wrong the error in the CI is implying that I must not update the ./helm-charts folder but the ./packaging/helm-charts instead. I'm fixing !

**Review by scholzj — COMMENTED** [review]
Thanks for the PR. It looks good, I just left some smaller comments. I also wonder if you can have a look at adding the ZooKeeper part to the `JmxST` system test. I do not think anyone in the core team really uses JMX, so having it covered in the ST would make sure it will keep working.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperatorTest.java
Comment (OuesFa) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperatorTest.java]: Thanks @scholzj  for your review!
Regarding this, I have some difficulties to know how to test correctly within `KafkaAssemblyOperatorNonParametrizedTest` . 
[I just pushed a version](https://github.com/strimzi/strimzi-kafka-operator/pull/5413/commits/e35eee6b80774b27799bcedf2a097d56ddf84d17) where I do a kind of dummy test that passes and where I describe a bit in comments how my target test and the main difficulty I found (not having a instance of ZookeeperCluster in order to call the generateJmxSecret() method)

Comment (OuesFa) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperatorTest.java]: It will be the same for the second test as well.

Comment (scholzj) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperatorTest.java]: Sorry, I was busy with the 0.25.0 release and some other work today. I will try to get to it tomorrow.

Comment (OuesFa) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperatorTest.java]: Of course, thank you!

Comment (OuesFa) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperatorTest.java]: I also tried something like this after the `Kafka` init 
```
        ResourceOperatorSupplier supplier = ResourceUtils.supplierWithMocks(false);
        SecretOperator secretOps = supplier.secretOperations;

        KafkaCluster kafkaCluster = KafkaCluster.fromCrd(Reconciliation.DUMMY_RECONCILIATION, kafka, VERSIONS);
        when(secretOps.getAsync(NAMESPACE, KafkaCluster.jmxSecretName(NAME))).thenReturn(
                Future.succeededFuture(kafkaCluster.generateJmxSecret(customAnnotations, customLabels))
        );

        KafkaAssemblyOperator op = new KafkaAssemblyOperator(vertx, new PlatformFeaturesAvailability(false, KubernetesVersion.V1_19), certManager, passwordGenerator,
                supplier, ResourceUtils.dummyClusterOperatorConfig(1L));
        Reconciliation reconciliation = new Reconciliation("test-trigger", Kafka.RESOURCE_KIND, NAMESPACE, NAME);

        Checkpoint async = context.checkpoint();

        op.new ReconciliationState(reconciliation, kafka).kafkaJmxSecret()
                .onComplete(context.succeeding(c -> context.verify(() -> {
                    assertThat(c.kafkaJmxSecret().succeeded(), is(Boolean.TRUE));

                    async.flag();
                })));
```
I get a null exception when calling `kafkaJmxSecret()`

Comment (OuesFa) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperatorTest.java]: My last commit contains a version that works. But it doesn't seem to follow your recommendations. 
Do you think it's ok ? 
If it's good I will add another test following the same pattern called `testUpdateZookeeperClusterAuthenticationTrue`

Comment (OuesFa) [reply cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperatorTest.java]: Much simpler approach indeed! Thanks 👍 
I just pushed the change.

**Review by scholzj — APPROVED** [review]

**Comment by OuesFa** [comment]
Regarding this test `testKafkaAndKafkaConnectWithJMX` that is failing with a NPE when launching regression.
It also fails when I run it locally, but when I put a breakpoint [on the faulty line](https://github.com/strimzi/strimzi-kafka-operator/blob/73a9efb537edebd4db0e52ca58707725cd932292/systemtest/src/main/java/io/strimzi/systemtest/resources/ResourceManager.java#L338) to debug in more detail, it ends up working, the test ends successfully, I think that it's related to the execution time.
I'm looking for a way to debug further the root cause.  
If you have any idea please let me know.

**Comment by scholzj** [comment]
@OuesFa I think you broke it. In your `KafkaAssemblyOperator.java` on line 1242, you use `KafkaCluster.jmxSecretName(name)` instead of `zkCluster.generateJmxSecret()`. This causes the Kafka JMX secret to be deleted.

**Comment by OuesFa** [comment]
> generateJmxSecret

**Comment by OuesFa** [comment]
> @OuesFa I think you broke it. In your `KafkaAssemblyOperator.java` on line 1242, you use `KafkaCluster.jmxSecretName(name)` instead of `zkCluster.generateJmxSecret()`. This causes the Kafka JMX secret to be deleted.

Nice catch thank you! Indeed I user the wrong class.

**Comment by scholzj** [comment]
@OuesFa Looks like that fixed it 👍

**Comment by OuesFa** [comment]
@scholzj I just rebased master, should we merge this today ?

**Comment by scholzj** [comment]
We need review / approval from another @strimzi/maintainers before we can merge it.

**Review by sknot-rh — APPROVED** [review]

**Comment by scholzj** [comment]
Thanks for the PR.

**Pull request merged**
